File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        if (nums.size() == 1) {
            return false;
        }

        if (nums.size() == 2) {
            return nums[1] == nums[0];
        }

        int sum = 0;

        for (int num : nums) {
            sum += num;
        }

        if (sum % 2 != 0) {
            return false;
        }

        int target = sum / 2;

        vector<int> dp(target + 1, 0);

        for (int num : nums) {
            for (int i = target; i >= num; i--) {
                dp[i] = max(dp[i], dp[i - num] + num);
                if (dp[i] == target) {
                return true;
                }
            }

            
        }

        return dp[target] == target;

        return dp[target];
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        totalSum = sum(nums)

        if totalSum % 2 != 0:
            return False

        target = totalSum // 2

        dp = [0] * (target + 1)

        for num in nums:
            for i in range(target, num - 1, -1):
                dp[i] = max(dp[i], num + dp[i - num])
            
            if dp[target] == target:
                return True

        return False

        